Alerta.io notifications

jrevillard opened the pull request:
Hello,

We are using Cruise Control in our infrastructure and as far as the alerting is concerned we are using [alerta.io](https://alerta.io/). I therefore created a notification plugin based on what was done for Slack. As far as we are using Kafka 2.6.X we did it in the 'migrate_to_kafka_2_5' branch but it should b straightforward  to do it in the other branches as well.

This plugin is now in production for us and works as expect since a week.

See: https://docs.alerta.io/en/latest/api/reference.html#create-an-alert

Best,
Jerome

P.S: it would be really good to also have notifications when an alert is closed (ether because this was a temporary failure (i.e: broker down) or because CC reconfigured the cluster.

efeg commented:
@jrevillard Thanks for the PR!
-- I will try to review it by next week.

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,250 @@
+        }
+    }
+
+    protected void sendAlertaMessage(AlertaMessage alertaMessage) throws IOException {
+        CloseableHttpClient client = HttpClients.createDefault();
+        HttpPost httpPost = new HttpPost(_alertaApiUrl + "/alert");
+        StringEntity entity = new StringEntity(alertaMessage.toString());
+        httpPost.setEntity(entity);
+        httpPost.setHeader("Authorization", "Key " + _alertaApiKey);
+        httpPost.setHeader("Accept", "application/json");
+        httpPost.setHeader("Content-type", "application/json");
+        try {
+            LOG.debug("Sending alert to Alerta.io : {}\nBody:\n{}", httpPost, alertaMessage);
+            CloseableHttpResponse httpResponse = client.execute(httpPost);
+            LOG.debug("Alerta.io response status: {}", httpResponse.getStatusLine().getStatusCode());
+        } finally {
+            client.close();
+        }
+    }
```
Given that we have similar code in `SlackSelfHealingNotifier`, would it make sense to 
1. Create a utility class under `notifier` package,
2. Extract the common logic to the utility class?

jrevillard replied:
Done

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,250 @@
+/*
```
Can we add documentation similar to `Configure-Slack-notifications.md` under `User Guide`?

jrevillard replied:
Done, I merged the 2 files

efeg commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifierTest.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Test
+    public void testSlackAlertWithNoApiKey() {
+        _notifier = new MockAlertaSelfHealingNotifier(MOCK_TIME);
+        _notifier._alertaApiUrl = "http://dummy.alerta.api";
+        _notifier.alert(FAILURES, false, 1L, KafkaAnomalyType.BROKER_FAILURE);
+        assertEquals(0, _notifier.getAlertaMessageList().size());
+    }
+
+    @Test
+    public void testSlackAlertWithDefaultOptions() {
+        _notifier = new MockAlertaSelfHealingNotifier(MOCK_TIME);
+        _notifier._alertaApiUrl = "http://dummy.alerta.api";
+        _notifier._alertaApiKey = "dummy-alerta-api-key";
+        _notifier.alert(FAILURES, false, 1L, KafkaAnomalyType.BROKER_FAILURE);
+        assertEquals(2, _notifier.getAlertaMessageList().size());
+    }
+
+    @Test
+    public void testSlackAlertWithEnvironment() throws ParseException {
+        _notifier = new MockAlertaSelfHealingNotifier(MOCK_TIME);
+        _notifier._alertaApiUrl = "http://dummy.alerta.api";
+        _notifier._alertaApiKey = "dummy-alerta-api-key";
+        _notifier._alertaEnvironment = "MyEnv";
+        long time = new SimpleDateFormat("yyyy-M-dd hh:mm:ssX").parse("2021-03-24 10:02:01Z").getTime();
+        _notifier.alert(FAILURES, false, time, KafkaAnomalyType.BROKER_FAILURE);
+        assertEquals(2, _notifier.getAlertaMessageList().size());
+        assertEquals("BROKER_FAILURE", _notifier.getAlertaMessageList().get(0).getEvent());
+        assertEquals("BROKER_FAILURE", _notifier.getAlertaMessageList().get(1).getEvent());
+        assertEquals(_notifier._alertaEnvironment, _notifier.getAlertaMessageList().get(0).getEnvironment());
+        assertEquals("2021-03-24T10:01:01.000Z", _notifier.getAlertaMessageList().get(0).getCreateTime());
+        assertEquals("2021-03-24T09:05:03.000Z", _notifier.getAlertaMessageList().get(1).getCreateTime());
+    }
```
Nit [Hardcoding]: Can we avoid hardcoding and instead use static final variable for test constants?

efeg reviewed (commented):
@jrevillard Thanks for the patch!
Left some comments.

efeg commented:
@jrevillard Once you get a chance to go over the feedback, would you please rebase on the latest `HEAD` in `migrate_to_kafka_2_5` branch? There has been some checkstyle rule additions that I suspect would affect this PR.

jrevillard commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,250 @@
+        }
+        final String localHostname = tmpLocalHostname;
+
+        DateTimeFormatter outFormatter = DateTimeFormatter
+            .ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSX") // millisecond precision
+            .withZone(ZoneId.of("UTC"));
```
Not possible as Alerta needs milliseconds precision.

efeg replied:
I see -- then can we
1. Add the following function to `CruiseControlUtils`:
```
  public static String utcDateFor(long timeMs, TemporalUnit unit) {
    return Instant.ofEpochMilli(timeMs).truncatedTo(unit).toString();
  }
```
2. Then use the response of this function in this class?

When called with `ChronoUnit.MILLIS`, this will return a response in ms precision, such as `2021-05-05T01:14:46.856Z`.

jrevillard commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,250 @@
+          case GOAL_VIOLATION:
+            GoalViolations goalViolations = (GoalViolations) anomaly;
+            Map<Boolean, List<String>> violations = goalViolations.violatedGoalsByFixability();
+            for (Entry<Boolean, List<String>> entry : violations.entrySet()) {
```
1. yes this is intentional, an the anomaly-id is always provided in the Tags.
2. This is available in the Raw Data

jrevillard commented:
Hi @efeg,

I did the rebase and fixed a couple of comments... but I'm still fighting with my Eclipse env because it does not accept your checkstyle file :-( so unfortunatly, I cannot check it properly.

Best.

jrevillard commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,250 @@
+        default:
+          return "warning";
+      }
+    }
```
I created an enum but I'm not sure if it's what you wanted...

jrevillard commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,250 @@
+              entry.getValue().forEach(goal -> { 
+                AlertaMessage alertaMessage = new AlertaMessage(localHostname, anomalyType.toString() + " - " + goal);
+                alertaMessage.setSeverity(severityFromPriority(anomalyType.priority()));
+                alertaMessage.setGroup("Performance");
```
Well, it's even not clear to me if I should do this classification.... I did the enum...

jrevillard replied:
By the way, here is an interresting documentaton I think: https://docs.alerta.io/en/latest/conventions.html

efeg commented:
> Hi @efeg,
> 
> I did the rebase and fixed a couple of comments... but I'm still fighting with my Eclipse env because it does not accept your checkstyle file :-( so unfortunatly, I cannot check it properly.
> 
> Best.

Thanks for the updates @jrevillard !
I see that you addressed style issues, but there seems to be a [test failure](https://app.circleci.com/pipelines/github/linkedin/cruise-control/996/workflows/04da3a99-2edb-447a-90da-300d72d7dec7/jobs/4138):

```
com.linkedin.kafka.cruisecontrol.detector.notifier.SlackSelfHealingNotifierTest > testSlackAlertWithDefaultOptions FAILED
    java.lang.AssertionError: expected:<1> but was:<0>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at com.linkedin.kafka.cruisecontrol.detector.notifier.SlackSelfHealingNotifierTest.testSlackAlertWithDefaultOptions(SlackSelfHealingNotifierTest.java:83)
```

Does this test pass locally for you?

jrevillard commented:
My bad, it was a bit lat ;-) .. this is fixed.

By the way, I have a 2 tests which are failing locally because my workstation is setup with French local:

```
com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest > testNullRecommendation FAILED
    org.junit.ComparisonFailure: expected:<...e cpu capacity (3200[.]00) as broker-0.> but was:<...e cpu capacity (3200[,]00) as broker-0.>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest.testNullRecommendation(ProvisionResponseTest.java:95)

com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest > testAggregate FAILED
    org.junit.ComparisonFailure: expected:<...e cpu capacity (3200[.00) as broker-0. [Recommender-Under-Provisioned] Add at least 6 brokers with the same cpu capacity (3200.]00) as broker-0.> but was:<...e cpu capacity (3200[,00) as broker-0. [Recommender-Under-Provisioned] Add at least 6 brokers with the same cpu capacity (3200,]00) as broker-0.>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest.testAggregate(ProvisionResponseTest.java:134)
```

Also, in core, I ave a test which fails locally: 

```
com.linkedin.cruisecontrol.CruiseControlUtilsTest > testUtcDate FAILED
    org.junit.ComparisonFailure: expected:<19[69-12-31T23:59:59]Z> but was:<19[70-01-01T00:00:00]Z>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at com.linkedin.cruisecontrol.CruiseControlUtilsTest.testUtcDate(CruiseControlUtilsTest.java:32)
```

jrevillard commented:
Hi @efeg , I think that I covered mostly all you remarks...

Best,
Jerome

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -102,112 +117,40 @@ public void alert(Anomaly anomaly, boolean autoFixTriggered, long selfHealingSta
         for (AlertaMessage alertaMessage : alertaMessages) {
           alertaMessage.setEnvironment(_alertaEnvironment);
-          alertaMessage.setService(Arrays.asList("cruise-control"));
+          alertaMessage.setService(Arrays.asList(ALERT_CRUISE_CONTROL));
```
Nit: Since it is a singleton, can we use `Collections.singletonList` rather than `Arrays.asList`?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -102,112 +117,40 @@ public void alert(Anomaly anomaly, boolean autoFixTriggered, long selfHealingSta
+          alertaMessage.setType(ALERT_CRUISE_CONTROL_ALARM);
           alertaMessage.setRawData(anomaly.toString());
-          alertaMessage.setTags(Arrays.asList("alarm_id:" + anomaly.anomalyId()));
+          alertaMessage.setTags(Arrays.asList(ALERT_ALARM_ID_TAG_KEY + ":" + anomaly.anomalyId()));
```
Nit: Since it is a singleton, can we use `Collections.singletonList` rather than `Arrays.asList`?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,282 @@
+      NotifierUtils.sendMessage(alertaMessage.toString(), _alertaApiUrl + "/alert", _alertaApiKey);
+    }
+
+    protected String severityFromPriority(int priority) {
```
Nit: Should we 
1. make this function `static`,
2. make it get `AnomalyType anomalyType` as input rather than `int priority`
3. move it to `NotifierUtils` with a JavaDoc,
4. rename the `AlertaAlertSeverity` to something more generic (e.g. `AlertSeverity`),
5. rename the function to something that represents its new role (e.g. `getAlertSeverity`)
5. make this function return the enum rather than the String representation of it?

This sets a standard mapping between `AnomalyType` and their severity, and it makes it possible to reuse this function in other notifiers if needed.

jrevillard replied:
Ok I did it even if I'm not convinced at all as this function will return an enum which is in any case Alerta specific even if its name change. Indeed, `AlertSeverity.toString()` must return something which is directly used by the Alerta API otherwise we simply introduce a useless intermediate enum.

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaAlertSeverity.java:
```diff
@@ -0,0 +1,9 @@
+ */
+
+package com.linkedin.kafka.cruisecontrol.detector.notifier;
+
```
Nit: Can we add JavaDoc to explain what this enum represents?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaAlertGroup.java:
```diff
@@ -0,0 +1,9 @@
+ */
+
+package com.linkedin.kafka.cruisecontrol.detector.notifier;
+
```
Nit: Can we add JavaDoc to explain what this enum represents?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,282 @@
+          tmpLocalHostname = InetAddress.getLocalHost().getCanonicalHostName();
+        } catch (UnknownHostException e1) {
+          LOG.warn("Unable to get the hostname of the Cruise Control server");
+          tmpLocalHostname = "Cruise Control";
```
Nit: Can we move the hardcoded value to a static variable?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/NotifierUtils.java:
```diff
@@ -0,0 +1,45 @@
+
+    private NotifierUtils() { }
+
+    /**
+     * Method used to send POST message to a specific URL (application/json)
+     */
```
Missing `@param` tags in JavaDoc.

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/NotifierUtils.java:
```diff
@@ -0,0 +1,45 @@
+        try {
+            LOG.debug("Sending alert to: {}\nBody:\n{}", httpPost, message);
+            CloseableHttpResponse httpResponse = client.execute(httpPost);
+            LOG.debug("Response status: {}", httpResponse.getStatusLine().getStatusCode());
```
Nit: Should we put this inside 
```
if (LOG.isDebugEnabled()) {
...
}
```

efeg commented on docs/wiki/User Guide/Configure-notifications.md:
```diff
@@ -15,3 +19,20 @@ User name to display in the slack notification message. (default: Cruise Control
+`alerta.self.healing.notifier.environment` (Optional)
+Environment attribute used to namespace the alert resource. (See: https://docs.alerta.io/en/latest/conventions.html#environments-services) (default: Alerta.io API default value)
+
+To enable Slack notification, set `anomaly.notifier.class=com.linkedin.kafka.cruisecontrol.detector.notifier.AlertaSelfHealingNotifier`
```
`Slack` -> `Alerta.io`

efeg reviewed (commented):
@jrevillard Thanks for the updates!
Made another pass ,and left more comments.

efeg commented:
> My bad, it was a bit late 😄 .. this is fixed.
> 

Thanks for the fixes!

> By the way, I have a 2 tests which are failing locally because my workstation is setup with French local:
> 
> ```
> com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest > testNullRecommendation FAILED
>     org.junit.ComparisonFailure: expected:<...e cpu capacity (3200[.]00) as broker-0.> but was:<...e cpu capacity (3200[,]00) as broker-0.>
>         at org.junit.Assert.assertEquals(Assert.java:117)
>         at org.junit.Assert.assertEquals(Assert.java:146)
>         at com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest.testNullRecommendation(ProvisionResponseTest.java:95)
> 
> com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest > testAggregate FAILED
>     org.junit.ComparisonFailure: expected:<...e cpu capacity (3200[.00) as broker-0. [Recommender-Under-Provisioned] Add at least 6 brokers with the same cpu capacity (3200.]00) as broker-0.> but was:<...e cpu capacity (3200[,00) as broker-0. [Recommender-Under-Provisioned] Add at least 6 brokers with the same cpu capacity (3200,]00) as broker-0.>
>         at org.junit.Assert.assertEquals(Assert.java:117)
>         at org.junit.Assert.assertEquals(Assert.java:146)
>         at com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest.testAggregate(ProvisionResponseTest.java:134)
> ```
> 


Ah, maybe we should update the tests to take local setup into account.

> Also, in core, I ave a test which fails locally:
> 
> ```
> com.linkedin.cruisecontrol.CruiseControlUtilsTest > testUtcDate FAILED
>     org.junit.ComparisonFailure: expected:<19[69-12-31T23:59:59]Z> but was:<19[70-01-01T00:00:00]Z>
>         at org.junit.Assert.assertEquals(Assert.java:117)
>         at org.junit.Assert.assertEquals(Assert.java:146)
>         at com.linkedin.cruisecontrol.CruiseControlUtilsTest.testUtcDate(CruiseControlUtilsTest.java:32)
> ```

The failure on `testUtcDate` would happen if you are not using `Java 11` for the development -- i.e. CC is on Java 11.

jrevillard commented:
I just implemented what you said here: https://github.com/linkedin/cruise-control/pull/1510#discussion_r626203618

But this does not work as the formatter will truncate milliseconds if it is equal to 0....

jrevillard commented:
Ok so I fixed the 

> > My bad, it was a bit late smile .. this is fixed.
> 
> Thanks for the fixes!
> 
> > By the way, I have a 2 tests which are failing locally because my workstation is setup with French local:
> > ```
> > com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest > testNullRecommendation FAILED
> >     org.junit.ComparisonFailure: expected:<...e cpu capacity (3200[.]00) as broker-0.> but was:<...e cpu capacity (3200[,]00) as broker-0.>
> >         at org.junit.Assert.assertEquals(Assert.java:117)
> >         at org.junit.Assert.assertEquals(Assert.java:146)
> >         at com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest.testNullRecommendation(ProvisionResponseTest.java:95)
> > 
> > com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest > testAggregate FAILED
> >     org.junit.ComparisonFailure: expected:<...e cpu capacity (3200[.00) as broker-0. [Recommender-Under-Provisioned] Add at least 6 brokers with the same cpu capacity (3200.]00) as broker-0.> but was:<...e cpu capacity (3200[,00) as broker-0. [Recommender-Under-Provisioned] Add at least 6 brokers with the same cpu capacity (3200,]00) as broker-0.>
> >         at org.junit.Assert.assertEquals(Assert.java:117)
> >         at org.junit.Assert.assertEquals(Assert.java:146)
> >         at com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponseTest.testAggregate(ProvisionResponseTest.java:134)
> > ```
> 
> Ah, maybe we should update the tests to take local setup into account.

I did it, tell me if it's ok for you. It works for me and in the CI.

> 
> > Also, in core, I ave a test which fails locally:
> > ```
> > com.linkedin.cruisecontrol.CruiseControlUtilsTest > testUtcDate FAILED
> >     org.junit.ComparisonFailure: expected:<19[69-12-31T23:59:59]Z> but was:<19[70-01-01T00:00:00]Z>
> >         at org.junit.Assert.assertEquals(Assert.java:117)
> >         at org.junit.Assert.assertEquals(Assert.java:146)
> >         at com.linkedin.cruisecontrol.CruiseControlUtilsTest.testUtcDate(CruiseControlUtilsTest.java:32)
> > ```
> 
> The failure on `testUtcDate` would happen if you are not using `Java 11` for the development -- i.e. CC is on Java 11.

Ok....


So I managed to fix this time representation issue... I think that I fixed your comment. Tell me if it's ok now.

Best,
Jerome

ykubota commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/notifier/AlertaSelfHealingNotifier.java:
```diff
@@ -0,0 +1,261 @@
+        AlertaMessage alertaMessage = new AlertaMessage(localHostname, anomalyType.toString() + " - " + maintenanceEvent.reasonSupplier());
+        alertaMessage.setSeverity(NotifierUtils.getAlertSeverity(anomalyType).toString());
+        alertaMessage.setGroup(AlertaAlertGroup.PERFORMANCE.toString());
+        alertaMessage.setCreateTime(CruiseControlUtils.utcDateFor(maintenanceEvent.detectionTimeMs(), 3, ChronoUnit.SECONDS));
```
Sorry for a question on old commit. 
I'm reading related implementation and I wondered why you need  precision for milliseconds even you truncate. I understood the reason you need milliseconds from this [comment](https://github.com/linkedin/cruise-control/pull/1510#issuecomment-832978240) but why you don't set ChronoUnit.MILLISECONDS on the 3rd parameter?
